text/0096-auto-batching.md

```diff
@@ -0,0 +1,67 @@
+- Title: Auto-Batching
+- Start Date: 2021-10-06
+- Specification PR: [#96](https://github.com/meilisearch/specifications/pull/96)
+- Discovery Issue: [#69](https://github.com/meilisearch/product/issues/69)
+
+# Auto-Batching
+
+## 1. Functional Specification
+
+### 1.1 Summary
+
+MeiliSearch can automatically group consecutive asynchronous `documentsAddition` or `documentsPartial` tasks for the same index into a single batch via an automatic batching mechanism.
+
+#### Summary Key points
+
+- A `batchUid` field is added on a fully qualified `task` API object.
+- An auto-batching mechanism groups consecutive `documentsAddition` or consecutive `documentsPartial` enqueued tasks for a similar index when a task is fetched from the FIFO task queue to be processed.
+
+### 1.2 Motivation
+
+We regularly tell users to batch their documents to speed up the indexing speed. We decided to integrate a simple scheduler to automatically batch consecutive tasks for an index to make it transparent for the users and enhance their quality of life using MeiliSearch.
+
+### 1.3 Explanations
+
+#### 1.3.1 Overall
+
+All `tasks` are linked to a specific `batchUid` field. This `batchUid` field permits to identify that several identical, consecutive tasks have been grouped in the same batch.
+
+Only consecutive `documentsAddition` and `documentsPartial` `tasks` for the same index can have a shared `batchUid` since the scheduler will be able to group them together. This means that all `tasks` concerning other operations will also be part of a `batchUid` having only one task. This could be extended in the future by enhancing the scheduler capabilities.
+
+This is to avoid having multiple incompressible computation time by grouping them together to increase performance when applied to a maximum number of documents.
+
+#### 1.3.2 Grouping tasks to a single batch
+
+The scheduling program that groups tasks within a single batch is triggered when an asynchronous `task` currently processed reaches a terminal state as `succeeded` or `failed`.
+
+In other words, when the next `task` should be picked from the FIFO task queue. The scheduler fetch and group all consecutive `documentsAddition` for a similar index in a batch until it encounters another task type or a similar task type but for a different index.
```

Comment:
Is it really more complex to continue to unpack events if you encounter a task from a different Index?

```suggestion
In other words, when the next `task` should be picked from the FIFO task queue. The scheduler fetch and group all consecutive `documentsAddition` for a similar index in a batch until it encounters another task type.
```

E.g.
Here is a task list (t: taskId, i: index, a: action (update or delete), b: batchUid)
```
last_task_id = 50;
last_batch_uid = 4;

t:51 i:movies 	a: update 	b: 5
t:52 i:movies 	a: update 	b: 5
t:53 i:shows 	a: delete 	b: ?
t:54 i:shows 	a: update 	b: ?
t:55 i:movies 	a: update 	b: 5
t:56 i:movies 	a: delete 	b: ?
t:57 i:shows 	a: update 	b: ?
t:58 i:shows 	a: update 	b: ?
t:59 i:movies 	a: update 	b: 5
t:60 i:movies 	a: delete 	b: ?

-- Next tick
last_task_id = 60;
last_batch_uid = 5;

t:53 i:shows 	a: delete 	b: 6
t:54 i:shows 	a: update 	b: ?
t:56 i:movies 	a: delete 	b: ?
t:57 i:shows 	a: update 	b: ?
t:58 i:shows 	a: update 	b: ?
t:60 i:movies 	a: delete 	b: ?
t:61 i:movies 	a: delete	b: ?

-- Next tick
last_task_id = 61;
last_batch_uid = 6;

t:54 i:shows 	a: update 	b: 7
t:56 i:movies 	a: delete 	b: ?
t:57 i:shows 	a: update 	b: 7
t:58 i:shows 	a: update 	b: 7
t:60 i:movies 	a: delete 	b: ?
t:61 i:movies 	a: delete	b: ?

-- Next tick
last_task_id = 61;
last_batch_uid = 7;

t:56 i:movies 	a: delete 	b: 8
t:60 i:movies 	a: delete 	b: 8
t:61 i:movies 	a: delete	b: 8

done.
```

Reply:
I don't even know if this is a good idea.

Reply:
Thanks, @qdequele  The goal is mainly to improve the process during the first massive document indexing that is not batched the right way by the user.

This is something we might want to do for the scheduler but it implies having a more intelligent and complex system. 

I'm thinking about compacting updates before adding documents for example. If the scheduler is not more intelligent, it seems to me that we end up with documents that are less up-to-date than they should be.

@MarinPostma your opinion will surely be much more enlightened than mine on this subject!

Reply:
Understood. Thanks @gmourier :)

Reply:
@qdequele We need to guarantee that updates are processed in the order the user provided them for an index, or at least that the the order we process them is equivalent. In you example, you process `[51, 52, 55, 59]` in the same batch, but `59 `happens after `56` which is a delete operation, and there is no reason to believe that processing 59 before 56 would yield the same result as the opposite order. e.g: `56` could delete docid 1, and `59` insert docid 1, it is easy to see that the order wuld matter in that case

Reply:
Indeed I made a mistake in my example that I corrected. :/

Reply:
I think we can solve this conversation. I agree that we can go to the simplest for the first iteration since the future evolutions will change the behavior through optimizations but not the API.